File: ass_editor.go
// ass_editor.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Event struct {
	Prefix   string
	StartMs  int
	EndMs    int
	Style    string
	Text     string
	Raw      string
}

type ASSParser struct {
	filename string
	header   []string
	events   []*Event
}

func NewASSParser(filename string) *ASSParser {
	return &ASSParser{filename: filename}
}

func (p *ASSParser) Parse() error {
	file, err := os.Open(p.filename)
	if err != nil {
		return err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	inEvents := false
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "[Events]") {
			inEvents = true
			p.header = append(p.header, line)
			continue
		}
		if !inEvents {
			p.header = append(p.header, line)
			continue
		}
		if strings.HasPrefix(line, "Format:") {
			p.header = append(p.header, line)
			continue
		}
		if strings.HasPrefix(line, "Dialogue:") || strings.HasPrefix(line, "Comment:") {
			ev, err := parseEvent(line)
			if err == nil {
				p.events = append(p.events, ev)
			}
		} else {
			p.header = append(p.header, line)
		}
	}
	return scanner.Err()
}

func parseEvent(line string) (*Event, error) {
	parts := strings.SplitN(line, ",", 10)
	if len(parts) < 10 {
		return nil, fmt.Errorf("invalid event line")
	}
	prefix := parts[0]
	startStr := strings.TrimSpace(parts[1])
	endStr := strings.TrimSpace(parts[2])
	style := strings.TrimSpace(parts[3])
	text := strings.TrimSpace(parts[9])
	startMs, err := timeToMs(startStr)
	if err != nil {
		return nil, err
	}
	endMs, err := timeToMs(endStr)
	if err != nil {
		return nil, err
	}
	return &Event{
		Prefix:  prefix,
		StartMs: startMs,
		EndMs:   endMs,
		Style:   style,
		Text:    text,
		Raw:     line,
	}, nil
}

func timeToMs(ts string) (int, error) {
	if ts == "" {
		return 0, nil
	}
	parts := strings.Split(ts, ":")
	if len(parts) != 3 {
		return 0, fmt.Errorf("invalid time format")
	}
	h, _ := strconv.Atoi(parts[0])
	m, _ := strconv.Atoi(parts[1])
	secPart := parts[2]
	var s, ms int
	if strings.Contains(secPart, ".") {
		sp := strings.SplitN(secPart, ".", 2)
		s, _ = strconv.Atoi(sp[0])
		millis := sp[1]
		if len(millis) > 3 {
			millis = millis[:3]
		}
		ms, _ = strconv.Atoi(millis)
	} else {
		s, _ = strconv.Atoi(secPart)
	}
	return (h*3600 + m*60 + s) * 1000 + ms, nil
}

func msToTime(ms int) string {
	sign := ""
	if ms < 0 {
		sign = "-"
		ms = -ms
	}
	h := ms / 3600000
	ms %= 3600000
	m := ms / 60000
	ms %= 60000
	s := ms / 1000
	ms %= 1000
	return fmt.Sprintf("%s%d:%02d:%02d.%03d", sign, h, m, s, ms)
}

func (p *ASSParser) ApplyShift(deltaMs int) {
	for _, ev := range p.events {
		if ev.StartMs+deltaMs < 0 {
			ev.StartMs = 0
		} else {
			ev.StartMs += deltaMs
		}
		if ev.EndMs+deltaMs < 0 {
			ev.EndMs = 0
		} else {
			ev.EndMs += deltaMs
		}
	}
}

func (p *ASSParser) ApplyReplace(old, new string) {
	re := regexp.MustCompile(old)
	for _, ev := range p.events {
		ev.Text = re.ReplaceAllString(ev.Text, new)
	}
}

func (p *ASSParser) ApplyStyle(newStyle string) {
	for _, ev := range p.events {
		ev.Style = newStyle
	}
}

func (p *ASSParser) Save(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	for _, h := range p.header {
		fmt.Fprintln(file, h)
	}
	for _, ev := range p.events {
		start := msToTime(ev.StartMs)
		end := msToTime(ev.EndMs)
		line := fmt.Sprintf("%s,%s,%s,%s,0,0,0,,%s", ev.Prefix, start, end, ev.Style, ev.Text)
		fmt.Fprintln(file, line)
	}
	return nil
}

func (p *ASSParser) ListEvents(color bool) {
	green, yellow, reset := "", "", ""
	if color {
		green = "\033[92m"
		yellow = "\033[93m"
		reset = "\033[0m"
	}
	for i, ev := range p.events {
		start := msToTime(ev.StartMs)
		end := msToTime(ev.EndMs)
		fmt.Printf("%s[%d]%s %s --> %s  %s%s%s: %s\n",
			green, i, reset, start, end, yellow, ev.Style, reset, ev.Text)
	}
}

func (p *ASSParser) ExportSrt(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	for i, ev := range p.events {
		start := strings.Replace(msToTime(ev.StartMs), ".", ",", -1)
		end := strings.Replace(msToTime(ev.EndMs), ".", ",", -1)
		fmt.Fprintf(file, "%d\n%s --> %s\n%s\n\n", i+1, start, end, ev.Text)
	}
	return nil
}

func main() {
	var (
		input     string
		output    string
		shift     int
		replace   string
		replaceTo string
		style     string
		list      bool
		exportSrt string
	)
	flag.StringVar(&input, "input", "", "Input ASS file")
	flag.StringVar(&output, "output", "", "Output file")
	flag.IntVar(&shift, "shift", 0, "Shift time in milliseconds")
	flag.StringVar(&replace, "replace", "", "Old text to replace (regex)")
	flag.StringVar(&replaceTo, "replace-to", "", "New text")
	flag.StringVar(&style, "style", "", "Change style name")
	flag.BoolVar(&list, "list", false, "List events")
	flag.StringVar(&exportSrt, "export-srt", "", "Export to SRT")
	flag.Parse()

	if input == "" {
		fmt.Fprintln(os.Stderr, "Error: --input required")
		os.Exit(1)
	}
	p := NewASSParser(input)
	if err := p.Parse(); err != nil {
		fmt.Fprintf(os.Stderr, "Parse error: %v\n", err)
		os.Exit(1)
	}

	if shift != 0 {
		p.ApplyShift(shift)
	}
	if replace != "" && replaceTo != "" {
		p.ApplyReplace(replace, replaceTo)
	}
	if style != "" {
		p.ApplyStyle(style)
	}
	if list {
		p.ListEvents(true)
	}
	if exportSrt != "" {
		if err := p.ExportSrt(exportSrt); err != nil {
			fmt.Fprintf(os.Stderr, "Export error: %v\n", err)
		}
	}
	if output != "" || shift != 0 || replace != "" || style != "" {
		out := output
		if out == "" {
			out = input
		}
		if err := p.Save(out); err != nil {
			fmt.Fprintf(os.Stderr, "Save error: %v\n", err)
		} else {
			fmt.Printf("Saved to %s\n", out)
		}
	}
}

File: ASSEditor.kt
// ASSEditor.kt
import java.io.File
import java.util.*
import kotlin.text.Regex

class ASSEditor {
    private val header = mutableListOf<String>()
    private val events = mutableListOf<Event>()

    data class Event(val prefix: String, var startMs: Int, var endMs: Int, var style: String, var text: String)

    fun parse(filename: String) {
        val lines = File(filename).readLines()
        var inEvents = false
        for (line in lines) {
            when {
                line.startsWith("[Events]") -> {
                    inEvents = true
                    header.add(line)
                }
                !inEvents -> header.add(line)
                line.startsWith("Format:") -> header.add(line)
                line.startsWith("Dialogue:") || line.startsWith("Comment:") -> {
                    events.add(parseEvent(line))
                }
                else -> header.add(line)
            }
        }
    }

    private fun parseEvent(line: String): Event {
        val parts = line.split(",", limit = 10)
        if (parts.size < 10) throw IllegalArgumentException("Invalid event")
        return Event(
            prefix = parts[0],
            startMs = timeToMs(parts[1].trim()),
            endMs = timeToMs(parts[2].trim()),
            style = parts[3].trim(),
            text = parts[9].trim()
        )
    }

    private fun timeToMs(ts: String): Int {
        if (ts.isEmpty()) return 0
        val parts = ts.split(":")
        if (parts.size != 3) return 0
        val h = parts[0].toInt()
        val m = parts[1].toInt()
        val secPart = parts[2]
        val (s, ms) = if (secPart.contains(".")) {
            val sp = secPart.split(".")
            val s = sp[0].toInt()
            val millis = sp[1].padEnd(3, '0').take(3)
            s to millis.toInt()
        } else {
            secPart.toInt() to 0
        }
        return (h * 3600 + m * 60 + s) * 1000 + ms
    }

    private fun msToTime(ms: Int): String {
        var msCopy = ms
        var sign = ""
        if (msCopy < 0) { sign = "-"; msCopy = -msCopy }
        val h = msCopy / 3600000
        msCopy %= 3600000
        val m = msCopy / 60000
        msCopy %= 60000
        val s = msCopy / 1000
        msCopy %= 1000
        return "$sign$h:${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}.${msCopy.toString().padStart(3, '0')}"
    }

    fun applyShift(deltaMs: Int) {
        for (e in events) {
            e.startMs = (e.startMs + deltaMs).coerceAtLeast(0)
            e.endMs = (e.endMs + deltaMs).coerceAtLeast(0)
        }
    }

    fun applyReplace(old: String, new: String) {
        val regex = Regex(old)
        for (e in events) {
            e.text = regex.replace(e.text, new)
        }
    }

    fun applyStyle(newStyle: String) {
        for (e in events) {
            e.style = newStyle
        }
    }

    fun save(filename: String) {
        val out = File(filename).printWriter()
        header.forEach { out.println(it) }
        events.forEach { e ->
            val start = msToTime(e.startMs)
            val end = msToTime(e.endMs)
            out.println("${e.prefix},$start,$end,${e.style},0,0,0,,${e.text}")
        }
        out.close()
        println("Saved to $filename")
    }

    fun listEvents(color: Boolean) {
        val g = if (color) "\u001B[92m" else ""
        val y = if (color) "\u001B[93m" else ""
        val r = if (color) "\u001B[0m" else ""
        events.forEachIndexed { i, e ->
            val start = msToTime(e.startMs)
            val end = msToTime(e.endMs)
            println("$g[$i]$r $start --> $end  $y${e.style}$r: ${e.text}")
        }
    }

    fun exportSrt(filename: String) {
        val out = File(filename).printWriter()
        events.forEachIndexed { i, e ->
            var start = msToTime(e.startMs).replace('.', ',')
            var end = msToTime(e.endMs).replace('.', ',')
            out.println(i + 1)
            out.println("$start --> $end")
            out.println(e.text)
            out.println()
        }
        out.close()
        println("Exported SRT to $filename")
    }
}

fun main(args: Array<String>) {
    val params = mutableMapOf<String, String>()
    var i = 0
    while (i < args.size) {
        if (args[i].startsWith("--")) {
            val key = args[i].substring(2)
            if (i + 1 < args.size && !args[i + 1].startsWith("--")) {
                params[key] = args[++i]
            } else {
                params[key] = ""
            }
        }
        i++
    }
    val input = params["input"] ?: run {
        System.err.println("Error: --input required")
        return
    }
    val editor = ASSEditor()
    editor.parse(input)

    params["shift"]?.let { editor.applyShift(it.toInt()) }
    if (params.containsKey("replace") && params.containsKey("replace-to")) {
        editor.applyReplace(params["replace"]!!, params["replace-to"]!!)
    }
    params["style"]?.let { editor.applyStyle(it) }
    if (params.containsKey("list")) {
        editor.listEvents(true)
    }
    params["export-srt"]?.let { editor.exportSrt(it) }
    if (params.containsKey("output") || params.containsKey("shift") || params.containsKey("replace") || params.containsKey("style")) {
        val output = params["output"] ?: input
        editor.save(output)
    }
}
